gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFARecoveryCodeEndpoint.java

```diff
@@ -178,4 +195,43 @@ private boolean failIfUserIsNotPresent(RoutingContext routingContext){
 
         return false;
     }
+
+    private Single<EnrolledFactorSecurity> generateRecoveryCode(io.gravitee.am.model.User endUser, Client client) {
+        final Factor recoveryFactor = getClientRecoveryFactor(client);
+        final FactorProvider recoveryFactorProvider = factorManager.get(recoveryFactor.getId());
+        final Map<String, Object> factorData = Map.of(
+                FactorContext.KEY_RECOVERY_FACTOR,
+                recoveryFactor, KEY_USER, endUser);
+        final FactorContext recoveryFactorCtx = new FactorContext(applicationContext, factorData);
+
+        return ((RecoveryFactor) recoveryFactorProvider).generateRecoveryCode(recoveryFactorCtx);
+    }
+
+    private Factor getClientRecoveryFactor(Client client) {
+        Set<String> factors = client.getFactors();
+        if(factors == null || factors.isEmpty()) {
+            return null;
+        }
+
+        final List<Factor> recoveryFactor = factors.stream()
+                .filter(f -> factorManager.get(f) != null)
+                .map(factorManager::getFactor)
+                .filter(f -> f.getFactorType().equals(FactorType.RECOVERY_CODE))
+                .collect(Collectors.toList());
+
+        return recoveryFactor.size() > 0 ? recoveryFactor.get(0) : null;
```

Comment:
Maybe return an Optional ?
```suggestion
        factors.stream()
                .filter(f -> factorManager.get(f) != null)
                .map(factorManager::getFactor)
                .filter(f -> f.getFactorType().equals(FactorType.RECOVERY_CODE))
                .findFirst()
```

Reply:
done